clients/python-wrapper/lakefs/reference.py

```diff
@@ -1,22 +1,179 @@
 """
 Module containing lakeFS reference implementation
 """
-from typing import Optional
+
+from __future__ import annotations
+from typing import Optional, Generator, NamedTuple, Literal
+
+import lakefs_sdk
 
 from lakefs.client import Client, DefaultClient
+from lakefs.exceptions import api_exception_handler
+from lakefs.object import Object
+from lakefs.object_manager import ObjectManager
+
+
+class Commit(NamedTuple):
+    """
+    NamedTuple representing a lakeFS commit's properties
+    """
+    id: str
+    parents: [str]
+    committer: str
+    message: str
+    creation_date: int
+    meta_range_id: str
+    metadata: Optional[dict[str, str]] = None
+
+
+class Change(NamedTuple):
+    """
+    NamedTuple representing a diff change between to refs in lakeFS
+    """
+    type: Literal["added", "removed", "changed", "conflict", "prefix_changed"]
+    path: str
+    path_type: Literal["common_prefix", "object"]
+    size_bytes: Optional[int]
 
 
 class Reference:
     """
     Class representing a reference in lakeFS.
     """
     _client: Client
-    repo_id: str
-    id: str
+    _repo_id: str
+    _id: str
+    _commit: Optional[Commit] = None
 
     def __init__(self, repo_id: str, ref_id: str, client: Optional[Client] = DefaultClient) -> None:
         self._client = client
-        self.repo_id = repo_id
-        self.id = ref_id
+        self._repo_id = repo_id
+        self._id = ref_id
+
+    @property
+    def repo_id(self) -> str:
+        """
+        Return the repository id for this reference
+        """
+        return self._repo_id
+
+    @property
+    def id(self) -> str:
+        """
+        Returns the reference id
+        """
+        return self._id
+
+    @property
+    def objects(self) -> ObjectManager:
+        """
+        Returns a ObjectManager object for this reference
+        """
+        # TODO: Implement
+
+    @staticmethod
+    def _get_generator(func, *args, max_amount: Optional[int] = None, **kwargs):
```

Comment:
Nothing about this is specific to references, could be reused for any other paginated resource.. I'm assuming that will happen in a later PR :)

Reply:
If the pattern returns - will take this function out of this scope :)